File: packages/android_intent_plus/lib/src/put_classes/put_parcelable.dart
import 'package:android_intent_plus/src/parcelable_classes/base/parcelable_base.dart';
import 'package:android_intent_plus/src/parcelable_classes/bundle.dart';
import 'package:android_intent_plus/src/put_classes/base/put_base.dart';

class PutParcelable extends PutBase<ParcelableBase> {
  PutParcelable({required String key, required ParcelableBase value})
      : super(key: key, value: value);

  @override
  String get javaClass => 'PutParcelable';

  factory PutParcelable.fromJson({
    required String key,
    required dynamic value,
  }) {
    final javaClass = value['javaClass'];
    if (javaClass == 'Bundle') {
      return PutParcelable(
        key: key,
        value: Bundle.fromJson(value['value']),
      );
    } else {
      throw UnimplementedError('Unknown PutParcelable, javaClass: $javaClass');
    }
  }

  @override
  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'key': key,
      'javaClass': javaClass,
      'value': value.toJson()
    };
  }
}

File: packages/android_intent_plus/android/src/main/java/dev/fluttercommunity/plus/androidintent/Bundle/PutClasses/PutParcelable.java
package dev.fluttercommunity.plus.androidintent.Bundle.PutClasses;

import android.os.Parcelable;

import org.json.JSONException;
import org.json.JSONObject;

import dev.fluttercommunity.plus.androidintent.Bundle.Constants;
import dev.fluttercommunity.plus.androidintent.Bundle.ParcelableClasses.Bundle;
import dev.fluttercommunity.plus.androidintent.Bundle.ParcelableClasses.base.ParcelableBase;
import dev.fluttercommunity.plus.androidintent.Bundle.PutClasses.base.PutBase;
import dev.fluttercommunity.plus.androidintent.GetType;

public class PutParcelable extends PutBase<ParcelableBase> {

  public static final String JAVA_CLASS = "PutParcelable";

  public PutParcelable(String key, ParcelableBase value) {
    super(key, JAVA_CLASS, value);
  }

  public static void convert(android.os.Bundle bundle, PutParcelable putParcelable) {
    bundle.putParcelable(putParcelable.key, convertToParcelable(putParcelable.value));
  }

  public static PutParcelable fromJson(JSONObject jsonObject) throws JSONException {
    final ParcelableBase parcelableBase = parcelableBaseFromJson(jsonObject.getJSONObject(Constants.VALUE));
    return new PutParcelable(jsonObject.getString(Constants.KEY), parcelableBase);
  }

  public static ParcelableBase parcelableBaseFromJson(JSONObject json) throws JSONException {
    //Add other parcelable classes here
    final String javaClass = json.getString(Constants.JAVA_CLASS);
    //noinspection SwitchStatementWithTooFewBranches
    switch (javaClass) {
      case Bundle.JAVA_CLASS:
        return Bundle.fromJson(json);
    }
    throw new RuntimeException(String.format("JavaClass (%s) not found (Json.parcelableBaseFromJson)", javaClass));
  }

  public static Parcelable convertToParcelable(ParcelableBase value) {
    if (value instanceof Bundle) {
      Bundle bundleParcelable = (Bundle) value;
      final android.os.Bundle bundle = new android.os.Bundle();
      for (PutBase<?> putBase : bundleParcelable.value) {
        PutBase.addToAndroidOsBundle(bundle, putBase);
      }
      return bundle;
    } else {
      throw new RuntimeException(GetType.name(value.getClass()));
    }
  }

  @Override
  public JSONObject toJson() throws JSONException {
    JSONObject jsonObject = new JSONObject();
    jsonObject.put("key", key);
    jsonObject.put("javaClass", JAVA_CLASS);
    jsonObject.put("value", ParcelableBase.toJson(value));
    return jsonObject;
  }
}
